machines: Implement adding virtual network interfaces to VM

skobyda opened the pull request:
Some part of dialog body is shared with "Virtual Network Interface Settings" edit dialog.

@garrett @andreasn what do you think about the design?

![Screenshot from 2019-09-04 16-28-36](https://user-images.githubusercontent.com/42733240/64265981-1fc89980-cf34-11e9-8882-4ffb415e2bf0.png)
![Screenshot from 2019-09-04 16-28-48](https://user-images.githubusercontent.com/42733240/64266004-2820d480-cf34-11e9-8882-4f7c0f07f06a.png)
![Screenshot from 2019-09-04 16-29-20](https://user-images.githubusercontent.com/42733240/64266012-2c4cf200-cf34-11e9-9125-c346569ff3a5.png)
**When VM is running, option to persistently/non-persistently attach device is visible:**
![Screenshot from 2019-09-04 16-29-35](https://user-images.githubusercontent.com/42733240/64266016-2f47e280-cf34-11e9-8baa-e304d0026a75.png)

andreasn commented on pkg/machines/components/nicAdd.jsx:
```diff
@@ -0,0 +1,242 @@
+                    type='checkbox'
+                    checked={dialogValues.setNetworkMac}
+                    onChange={e => onValueChanged('setNetworkMac', !dialogValues.setNetworkMac)} />
+                {_("Set MAC address")}
```
Is this a child of `Model`? The label makes it look like it does.
Maybe better to put it next to the `Mac Address` label to make the grouping more clear.
Also, what happens if I do not chck this? It doesn't get a mac address?

skobyda replied:
No MAC address is not child of Model and not related to it.

MAC address is optional input. So if you choose not to set it, libvirt will autogenerate MAC address for you.
But there are use cases for allowing user to set MAC address (e.g. he might have DNS configuration with IPs statically assigned to some pre-existing MACs)

skobyda replied:
You mean putting it next to `MAC Address` label similar to how `Custom mount options` is next to input fields in **New NFS Mount** in Storage page?

andreasn replied:
Yes, like this:
![64266012-2c4cf200-cf34-11e9-9125-c346569ff3a5](https://user-images.githubusercontent.com/209895/64324450-b3e34100-cfc6-11e9-8942-68836d973ecb.png)

skobyda replied:
Done. Repushed.

andreasn commented on pkg/machines/components/nicAdd.jsx:
```diff
@@ -0,0 +1,242 @@
+
+            {dialogValues.setNetworkMac && <>
+                <label className='control-label' htmlFor={`${idPrefix}-mac`}>
+                    {_("Mac Address")}
```
I would prefer if this was MAC in upper case, to keep the spelling consistent, and avoid any confusion with the Macintosh computer.

andreasn commented on pkg/machines/components/nicBody.jsx:
```diff
@@ -0,0 +1,184 @@
+            {["network", "direct"].includes(dialogValues.networkType) && (
+                <>
+                    <label className='control-label' htmlFor={`${idPrefix}-select-source`}>
+                        {_("Source")}
```
How come source is to the right rather than below `Interface name`?
I don't have a strong opinion either way, just wanted to verify if there was any reason for it.

skobyda replied:
> `Interface name` ?
Do you mean `Interface type` ? The reason is that Source is heavily dependent on Interface type (child element if you want).
Also this reuses the component from Editing the Network Interface, where Source is also next to Network Type.
So alternatively we could show hierarchy by putting Source underneath the Interface name and shifting it a bit right (same as we did with hierarchy in Creation of Virtual Networks: https://github.com/cockpit-project/cockpit/pull/12266#issuecomment-510111787)

andreasn commented:
So what about something like this?
* Indicates the hierarchy between `Interface type` and `Source`.
* Makes it clear what happens if you choose to not manually enter a MAC address
* Puts the MAC address input on the same line as the control, avoiding dialog resize when that radiobutton is selected.

skobyda commented:
I like that mockup @andreasn . I implemented the changes and repushed.
Here is how it looks now (just small nickpick, I changed sentence of "Automatically generate" to "Generate automatically", to keep consistenci with "Set manually" which goes like "verb adverb".

So now both "Generate automatically" and "Set manually" are both "verb adverb". Hope you agree.

**If you choose `Generate automatically`, the input field for manual setting gets disabled**
![Screenshot from 2019-09-05 13-58-32](https://user-images.githubusercontent.com/42733240/64340124-b6a15e80-cfe5-11e9-9cfe-1883f2c4260c.png)

![Screenshot from 2019-09-05 13-58-42](https://user-images.githubusercontent.com/42733240/64340166-cd47b580-cfe5-11e9-8b5d-71b1ce104952.png)

The change in hierarchly also affects the Network Interface Edit dialog. Because we do not have 2 inputs next to each other, now I could resize the original wide dialog into default size.
![Screenshot from 2019-09-05 13-53-33](https://user-images.githubusercontent.com/42733240/64340260-fff1ae00-cfe5-11e9-99ed-08079b9d7c62.png)

![Screenshot from 2019-09-05 13-53-33](https://user-images.githubusercontent.com/42733240/64340171-d0db3c80-cfe5-11e9-8798-ccbd9fac222c.png)

andreasn commented:
> So now both Generate automatically and Set manually are both verb followed by adverb. Hope you agree.

Oh yes, great catch. Fully agree!

andreasn commented:
> The change in hierarchly also affects the Network Interface Edit dialog. Because we do not have 2 inputs next to each other, now I could resize the original wide dialog into default size.

Seems the Edit dialog still spells it `Mac` in lowercase, but I don't think that's blocking this PR. I can file a separate PR about that.

martinpitt commented:
This conflicts with master now. Can you please rebase?

Labelling release-note.

KKoukiou commented on pkg/machines/components/nicAdd.jsx:
```diff
@@ -0,0 +1,255 @@
+    return (
+        <>
+            <>
+                <label className='control-label' htmlFor={`${idPrefix}-set-mac`}>
```
Do you really want the htmlFor to be *-set-mac here?
This will move the focus to the Manual MAC settings if clicking on the label?
I think we should rather put the ID for the auto-generated setting here.

KKoukiou commented on pkg/machines/components/nicAdd.jsx:
```diff
@@ -0,0 +1,255 @@
+                let availableSources;
+                if (value == 'network')
+                    availableSources = this.props.networks.map(network => network.name);
+                else if (['direct'].includes(value))
```
Please use consistent way to compare with the string. 
Like if (value == 'network') use
else if (value == 'direct')

KKoukiou commented on pkg/machines/components/nicAdd.jsx:
```diff
@@ -0,0 +1,255 @@
+                    addDisabled = true;
+                }
+            }
+            this.setState({ addDisabled: addDisabled });
```
this.setState({ addDisabled })

KKoukiou commented on pkg/machines/components/nicBody.jsx:
```diff
@@ -0,0 +1,186 @@
+    );
+};
+
+export const NICBody = ({ idPrefix, dialogValues, onValueChanged, networks, interfaces, nodeDevices, connectionName, osTypeArch, osTypeMachine }) => {
```
I don't see why this component should be independent.
Just export the NetworkTypeAndSourceRow and the NetworkModelRow and use them from the Creation for NIC component.

KKoukiou commented on pkg/machines/components/nicAdd.jsx:
```diff
@@ -0,0 +1,255 @@
+import React from 'react';
+import cockpit from 'cockpit';
+import PropTypes from 'prop-types';
+import {
```
Same line better? It's not that long.

KKoukiou commented on pkg/machines/components/nicAdd.jsx:
```diff
@@ -0,0 +1,255 @@
+import { getNetworkDevices } from '../helpers.js';
+import { NICBody } from './nicBody.jsx';
+import {
+    getVm
```
Same line here please?

KKoukiou commented on pkg/machines/components/nicBody.jsx:
```diff
@@ -0,0 +1,186 @@
+        { name: 'rtl8139', desc: 'PCI, legacy' }];
+    const defaultModelType = dialogValues.networkModel;
+
+    if (osTypeArch == 'ppc64' && osTypeMachine == 'pseries') {
```
No need for curly brackets in single liners.

KKoukiou commented on pkg/machines/components/nicBody.jsx:
```diff
@@ -0,0 +1,186 @@
+    const defaultModelType = dialogValues.networkModel;
+
+    if (osTypeArch == 'ppc64' && osTypeMachine == 'pseries') {
+        availableModelTypes.push('spapr-vlan');
```
Hmm, the format of the available Model Types is { name: NAME, desc: DESC }  - here you just push a string.

skobyda replied:
This is the stuff which I pulled from other component and did not changed. I have no idea, what description should go next to "spapr-vlan" since i don't know that model. 
So I changed a component to use the _desc_ value optionally.

skobyda replied:
All comments resolved.

KKoukiou commented on pkg/machines/components/nicBody.jsx:
```diff
@@ -0,0 +1,186 @@
+    let networkSourceEnabled = true;
+    const networkDevices = getNetworkDevices(connectionName, nodeDevices, interfaces);
+
+    if (connectionName !== 'session')
```
For multiliners we tend to use curly brackets.

skobyda commented:
It seems there are no machines related failures (apart from selenium-edge :( )
